cmd/kubeadm/app/cmd/certs_test.go

```diff
@@ -482,3 +485,170 @@ kubernetesVersion: %s`,
+	}
+	for _, kubeConfig := range kubeConfigs {
+		if err := kubeconfigphase.CreateKubeConfigFile(kubeConfig, kdir, cfg); err != nil {
+			t.Fatalf("couldn't create kubeconfig %q: %v", kubeConfig, err)
+		}
+	}
+
+	// A minimal kubeadm config with just enough values to avoid triggering
+	// auto-detection of config values at runtime.
+	var kubeadmConfig = fmt.Sprintf(`
+apiVersion: %[1]s
+kind: ClusterConfiguration
+certificatesDir: %s
+kubernetesVersion: %s`,
+		kubeadmapiv1.SchemeGroupVersion.String(),
+		cfg.CertificatesDir,
+		kubeadmconstants.MinimumControlPlaneVersion.String())
+
+	customConfigPath := kdir + "/kubeadm.conf"
+	f, err := os.Create(customConfigPath)
+	require.NoError(t, err)
+	_, err = f.Write([]byte(kubeadmConfig))
+	require.NoError(t, err)
+
+	commArgs := []string{
+		fmt.Sprintf("--cert-dir=%s", cfg.CertificatesDir),
+		fmt.Sprintf("--config=%s", customConfigPath),
+	}
+
+	type testCase struct {
+		name           string
+		output         string
+		brokenCertName string
+	}
+
+	var runTestCase = func(t *testing.T, tc testCase) {
+		var output bytes.Buffer
+		cmd := newCmdCertsExpiration(&output, kdir)
+		cmd.SetArgs(commArgs)
+		if tc.output != "" {
+			args := append(commArgs, fmt.Sprintf("-o=%s", tc.output))
+			cmd.SetArgs(args)
+		}
+		require.NoError(t, cmd.Execute())
+
+		switch tc.output {
+		case "json":
+			var info outputapiv1alpha3.CertificateExpirationInfo
+			require.NoError(t, json.Unmarshal(output.Bytes(), &info))
+			assert.Len(t, info.Certificates, len(kubeadmCerts)+len(kubeConfigs))
+			assert.Len(t, info.CertificateAuthorities, len(caCerts))
+			for _, cert := range info.Certificates {
+				if tc.brokenCertName == cert.Name {
+					assert.True(t, cert.Missing, "expected certificate to be missing")
+				} else {
+					assert.False(t, cert.Missing, "expected certificate to be present")
+				}
+			}
+		default:
+			outputStr := output.String()
+			if tc.brokenCertName != "" {
+				assert.Contains(t, outputStr, "!MISSING!")
+			} else {
+				assert.NotContains(t, outputStr, "!MISSING!")
+			}
+
+			var lines []string
+			for _, line := range strings.SplitAfter(outputStr, "\n") {
+				if strings.TrimSpace(line) != "" {
+					lines = append(lines, line)
+				}
+			}
+			// 2 lines for the column headers.
+			expectedLineCount := len(caCerts) + len(kubeadmCerts) + len(kubeConfigs) + 2
+			assert.Lenf(t, lines, expectedLineCount, "expected %d non-blank lines in output", expectedLineCount)
+		}
+	}
+
+	testCases := []testCase{
+		{
+			name:   "print columns and no missing certs",
+			output: "",
+		},
+		{
+			name:   "print json and no missing certs",
+			output: "json",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			runTestCase(t, tc)
+		})
+	}
+
+	brokenCertName := kubeadmconstants.APIServerCertAndKeyBaseName
+	brokenCertPath, _ := pkiutil.PathsForCertAndKey(cfg.CertificatesDir, brokenCertName)
+	os.Remove(brokenCertPath)
+
+	brokenTestCases := []testCase{
```

Comment:
i think we can combine all test cases in the same testCases list.
this is how we do it in the rest of kubeadm.